File: Problems/20. Valid Parentheses/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        table = {
            "(":")",
            "[":"]",
            "{":"}"};
        seen = [];

        for symbol in s:
            if symbol in table:
                seen.append(symbol);
            else:
                if not seen: return False;
                if symbol!=table[seen.pop()]: return False;
        
        return not bool(seen);

File: Problems/20. Valid Parentheses/20. Valid Parentheses.java
import java.util.HashMap;
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> table = new HashMap<>();
        Stack<Character> stack = new Stack<>();
        table.put('[', ']');
        table.put('(', ')');
        table.put('{', '}');
        
        for (char ch: s.toCharArray()) {
            if (table.containsKey(ch)) stack.push(ch);
            else {
                if (stack.empty()) return false;
                if (ch != table.get(stack.pop())) return false;
            }
        }

        return stack.empty();
    }
}